icicle v 1.0.0 docs (#10)
ingonyama-zk/developer-docs: docs/icicle/integrations.md

# Integrations

ICICLE has been used by companies like [Celer Network](https://github.com/celer-network), [Gnark](https://github.com/Consensys/gnark), [EZKL](https://blog.ezkl.xyz/post/acceleration/) and others to accelerate their ZK proving pipeline.

Many of these integrations have been a collaboration between Ingonyama and the integrating company. We have learned a lot about designing GPU based ZK provers.  

> If your interested in understanding these integrations better and learning how you can use ICICLE to accelerating your existing ZK proving pipeline this is the place for you.

Reviewer: ```suggestion
If you're interested in understanding these integrations better or learning how you can use ICICLE to accelerate your existing ZK proving pipeline this is the place for you.
```

Lets illustrate an ICICLE integration, so you can understand the core API and design overview of ICICLE.

> ![ICICLE architecture](../../static/img/architecture-highlight.png)

Reviewer: The image shows that only Rust and Golang APIs are highlevel APIs, but the c++ wrappers/API is also a highlevel api in some sense. For an application that is in c/c++, there is no need to use Rust/Golang APIs, it can use the c++ directly.

> Engineers usually use a cryptography library to implement their ZK protocols. These libraries implement efficient primitives which are used as building blocks for the protocol. ICICLE is such a library the difference is that ICICLE is designed from the start to run on GPUs. The Rust and Golang APIs abstract away all low level CUDA details, our goal was to allow developers with no GPU experience quickly get started with ICICLE. 

Reviewer: ```suggestion
Engineers usually use a cryptography library to implement their ZK protocols. These libraries implement efficient primitives which are used as building blocks for the protocol; ICICLE is such a library. The difference is that ICICLE is designed from the start to run on GPUs; the Rust and Golang APIs abstract away all low level CUDA details. Our goal was to allow developers with no GPU experience to quickly get started with ICICLE. 
```

A developer may use ICICLE with two main approaches in mind.

1. Drop in replacement approach.
2. Full GPU replacement approach.

> The first approach is best suited for teams aiming to quickly implement a GPU version of their Prover. This method predominantly involves substituting the core methods with ICICLE's methods. While this approach can yield significant performance gains it has many draw backs, implementing memory optimizations is limited and optimizing the protocol itself to be more suited for GPU is also limited. This approach is a great way to begin using ICICLE, however for someone who is looking to seriously adopt GPU acceleration we would recommend a more rigorous approach.

Reviewer: `While this approach can yield significant performance gains it has many draw backs, implementing memory optimizations is limited and optimizing the protocol itself to be more suited for GPU is also limited.`

I think we can reword this sentence to flow a bit better

Author: ok attempted to make it better

## A primer to building your own integrations

Lets illustrate an ICICLE integration, so you can understand the core API and design overview of ICICLE.

![ICICLE architecture](../../static/img/architecture-high-level.png)

Engineers usually use a cryptography library to implement their ZK protocols. These libraries implement efficient primitives which are used as building blocks for the protocol; ICICLE is such a library. The difference is that ICICLE is designed from the start to run on GPUs; the Rust and Golang APIs abstract away all low level CUDA details. Our goal was to allow developers with no GPU experience to quickly get started with ICICLE. 

A developer may use ICICLE with two main approaches in mind.

1. Drop in replacement approach.
2. Full GPU replacement approach.

The first approach for GPU-accelerating your Prover with ICICLE is quick to implement, but it has limitations, such as reduced memory optimization and limited protocol tuning for GPUs. It's a solid starting point, but those committed to fully leveraging GPU acceleration should consider a more comprehensive approach.

> A full GPU replacement means preforming the entire ZK proof on the GPU. This approach will reduce latency to a minimum and requires you to change the way you implement the protocol to be more GPU friendly. This approach will take full advantage of GPU acceleration. Redesigning your prover this way may take more engineering effort but we promise you that its worth it!

Reviewer: ```suggestion
A full GPU replacement means performing the entire ZK proof on the GPU. This approach will reduce latency to a minimum and requires you to change the way you implement the protocol to be more GPU friendly. This approach will take full advantage of GPU acceleration. Redesigning your prover this way may take more engineering effort but we promise you that its worth it!
```

## Using ICICLE integrations

> Here we cover how a developer can run his existing circuits on ICICLE integrated provers.

Reviewer: ```suggestion
Here we cover how a developer can run existing circuits on ICICLE integrated provers.
```

### Gnark

[Gnark](https://github.com/Consensys/gnark) officially supports GPU proving with ICICLE. Currently only Groth16 on curve `BN254` is supported. This means that if you are currently using Gnark to write your circuits you can enjoy GPU acceleration without making many changes.

> To use GPUs, add the icicle buildtag to your build/run commands, e.g. `go run -tags=icicle main.go.`

Reviewer: ```suggestion
To use GPUs, add the `icicle` buildtag to your build/run commands, e.g. `go run -tags=icicle main.go.`
```
